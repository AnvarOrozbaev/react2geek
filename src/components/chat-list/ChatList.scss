.chats-wrapper{
  display: flex;
  flex-direction: column;
}
$primary:   #864f12;
.add-chat{
&__input{
  border-radius: 10px;
  border: 1px solid $primary;
  border-right: none;
  border-top-right-radius: 0 ;
  border-bottom-right-radius: 0 ;
  outline: none;
  height: 1.2em;
}
&__btn{
  border-radius: 10px;
  border: 1px solid $primary;
  border-top-left-radius: 0 ;
  border-bottom-left-radius: 0 ;
  background-color: $primary;
  color: #ffffff;
  &:hover{
    background-color: #ffffff;
    color: $primary;
  }
  
}
}
