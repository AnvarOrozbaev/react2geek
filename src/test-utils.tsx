import React, { ReactElement } from 'react';
import { render as rtlRender } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import chatsReducer from './pages/chats/chatsSlice';
import profileReducer from './pages/profile/profileSlice';

type Props = {
  children?: JSX.Element;
};
function render(
  ui: ReactElement,
  {
    store = configureStore({
      reducer: {
        chatsState: chatsReducer,
        profileState: profileReducer,
      },
    }),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }: Props) {
    return <Provider store={store}>{children}</Provider>;
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

export * from '@testing-library/react';
export { render };
