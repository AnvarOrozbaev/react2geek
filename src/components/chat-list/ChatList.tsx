import React, { FC, useMemo } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useNavigate } from 'react-router'
import { ChatListItem } from '../chat-list-item/ChatListItem';
import { ChatAddForm } from '../chat-add-form/ChatAddForm';
import {
  deleteChat,
  setSelectedId,
  addChat,
  selectChats,
  selectSelectedId,
} from '../../pages/chats/chatsSlice';
import List from '@mui/material/List';
import { ChatProps, Entries, ChatsType } from '../types';
import './ChatList.scss';
import { nanoid } from '@reduxjs/toolkit';

export const ChatList: FC = () => {
  const chats = useSelector(selectChats);
  const navigate = useNavigate();
  const dispatch = useDispatch();
  const handleAddChat = (name: string) => {
    const id = nanoid();
    dispatch(addChat({ name, id }));
  };
  const entries = useMemo<Entries<ChatsType>>(
    () => Object.entries(chats),
    [chats]
  );
  const selectedId = useSelector(selectSelectedId);
  const handleDeleteChat = (id: string) => {
    dispatch(deleteChat(id));
    dispatch(setSelectedId(''));
    navigate('/chats');
  };
  const handleListItemClick = (id: string) => {
    dispatch(setSelectedId(id));
  };
  return (
    <div className="chats-wrapper">
      <ChatAddForm handleAddChat={handleAddChat} />
      <List sx={{ maxWidth: 150 }}>
        {entries.map((chat) => {
          const chatProps: ChatProps = { name: chat[1].name, id: chat[0] };
          return (
            <ChatListItem
              key={chat[0]}
              chat={chatProps}
              selectedId={selectedId}
              handleListItemClick={handleListItemClick}
              deleteChat={handleDeleteChat}
            />
          );
        })}
      </List>
    </div>
  );
};
