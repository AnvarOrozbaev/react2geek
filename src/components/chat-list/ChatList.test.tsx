import React from 'react';
import { render, fireEvent } from '../../test-utils';
import { ChatList } from './ChatList';
import { BrowserRouter } from 'react-router-dom';
import '@testing-library/jest-dom';

describe('ChatList', () => {
  const mockAddChat = jest.fn();
  it('render component', () => {
    render(
      <BrowserRouter>
        <ChatList />
      </BrowserRouter>
    );
  });
  it('render with snapshot', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <ChatList />
      </BrowserRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it('form behavior when input is empty', async () => {
    const { getByTestId, getByPlaceholderText } = render(
      <BrowserRouter>
        <ChatList />
      </BrowserRouter>
    );
    const btn = getByTestId('add-chat__btn');
    const input = getByPlaceholderText(/название чата/i);

    fireEvent.change(input, { target: { value: '' } });
    fireEvent.click(btn);
    expect(mockAddChat).toHaveBeenCalledTimes(0);
  });
});
