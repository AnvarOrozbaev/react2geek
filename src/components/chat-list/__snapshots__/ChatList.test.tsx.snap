// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatList render with snapshot 1`] = `
<DocumentFragment>
  <div
    class="chats-wrapper"
  >
    <form
      class="add-chat"
    >
      <input
        class="add-chat__input"
        placeholder="название чата"
        value=""
      />
      <input
        class="add-chat__btn"
        data-testid="add-chat__btn"
        type="submit"
        value="+"
      />
    </form>
    <ul
      class="MuiList-root MuiList-padding css-1uvrni2-MuiList-root"
    />
  </div>
</DocumentFragment>
`;
